### PR TITLE
Flashlights use batteries

### PR DESCRIPTION
**Pull request by etherware-novice**
## About The Pull Request
Flashlights use an internal battery, and the light starts dying at around 500kw left, before dimming further at 200kw and cutting out at 0

## Why It's Good For The Game
Flashlights are so, so easy to get and discount every other light emitter (besides maybe the hardhat?)

also its setting up my next pr stay tuned :)

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

image

</details>

## Changelog
:cl:
balance: Nanotrasen has stopped using infinite cells for their flashlights, downgrading to basic (flashlights have internal cells now)
/:cl:

### CODE REVIEW

File: code/game/objects/items/devices/flashlight.dm [hunk 1]
```diff
@@ -17,18 +25,83 @@
...
 	if(icon_state == "[initial(icon_state)]-on")
 		on = TRUE
 	update_brightness()
 
+/obj/item/flashlight/process(delta_time)
+	if(!uses_battery)
+		STOP_PROCESSING(SSobj, src)
+		return
+	if(int_battery)
+		int_battery.charge = max(int_battery.charge - (battery_use_rate * delta_time), 0)
```
Comment (Dejaku51) [inline code/game/objects/items/devices/flashlight.dm]: maybe use ```int_battery.use(battery_use_rate * delta_time)``` instead?

Comment (etherware-novice) [reply code/game/objects/items/devices/flashlight.dm]: oo i didnt know that was a proc oke

File: code/game/objects/items/devices/flashlight.dm [hunk 2]
```diff
@@ -107,6 +118,13 @@
 		A.UpdateButtonIcon()
 	return 1
 
+/obj/item/flashlight/screwdriver_act(mob/user, obj/item/I)
+	if(int_battery)
+		to_chat(user, "<span class='notice'>You remove the battery from \the [src]</span>")
+		user.put_in_hands(int_battery)
+		int_battery = null
```
Comment (Dejaku51) [inline code/game/objects/items/devices/flashlight.dm:136]: you can remove it but can you insert it back?

**Comment by Dejaku51** [comment]
attackby, not attack_by

**Comment by EvilDragonfiend** [comment]
You should be able to pull out the battery without screwdriver. Alt click to pull out the battery is enough.

File: code/game/objects/items/devices/flashlight.dm [hunk 3]
```diff
@@ -17,18 +25,84 @@
...
 	var/on = FALSE
+	/// uses internal battery
+	var/uses_battery = TRUE
+	var/obj/item/stock_parts/cell/int_battery
+	/// how much charge this battery starts with on init, does literally nothing else
+	var/start_battery_charge
+	/// how fast the battery is used
+	var/battery_use_rate = 0.5
+	/// lights power level
+	var/battery_power_amt
```
Comment (itsmeow) [inline code/game/objects/items/devices/flashlight.dm:36]: Change this to `flashlight_power_level`

File: code/game/objects/items/devices/flashlight.dm [hunk 4]
```diff
@@ -17,18 +25,84 @@
 	light_power = 1
 	light_on = FALSE
 	var/on = FALSE
+	/// uses internal battery
+	var/uses_battery = TRUE
+	var/obj/item/stock_parts/cell/int_battery
+	/// how much charge this battery starts with on init, does literally nothing else
+	var/start_battery_charge
```
Comment (itsmeow) [inline code/game/objects/items/devices/flashlight.dm:32]: initial_charge

File: code/game/objects/items/devices/flashlight.dm [hunk 5]
```diff
@@ -17,18 +25,84 @@
...
+	/// how fast the battery is used
+	var/battery_use_rate = 0.5
+	/// lights power level
+	var/battery_power_amt
+
+/obj/item/flashlight/examine(mob/user)
+	. = ..()
+	if(uses_battery)
+		if(istype(int_battery))
+			. += "The flashlight is at [round(int_battery.percent() )]%."
```
Comment (itsmeow) [inline code/game/objects/items/devices/flashlight.dm:42]: random space + no span

File: code/game/objects/items/devices/flashlight.dm [hunk 6]
```diff
@@ -17,18 +25,84 @@
...
+			. += "<span class='notice'>Alt-click to remove cell.</span>"
+		else
+			. += "There doesn't seem to be a cell inserted."
```
Comment (itsmeow) [inline code/game/objects/items/devices/flashlight.dm:45]: no span

File: code/game/objects/items/devices/flashlight.dm [hunk 7]
```diff
@@ -17,18 +25,84 @@
...
+				if(LIGHT_POWER_HIGH)
+					light_range = initial(light_range)
+				if(LIGHT_POWER_LOW)
+					light_range = initial(light_range) / 2
+				if(LIGHT_POWER_CRIT)
+					light_range = initial(light_range) / 4
+				else
+					to_chat(user, "\The [src] flickers weakly, but doesn't have enough power to turn on.")
+					on = FALSE
+					return .()
```
Comment (itsmeow) [inline code/game/objects/items/devices/flashlight.dm:101]: what the hell is `.()`

just return, there is no parent proc

File: code/game/objects/items/devices/flashlight.dm [hunk 8]
```diff
@@ -45,6 +119,23 @@
...
 
+/obj/item/flashlight/attackby(obj/item/I, mob/user)
+	if(istype(I, /obj/item/stock_parts/cell))
+		if(uses_battery)
+			if(int_battery)
+				to_chat(user, "<span class='warning'>There's already a cell inside!</span>")
+				return
+			int_battery = I
+			I.forceMove(src)
+			to_chat(user, "<span class='notice'>You plug the cell into \the [src].")
```
Comment (itsmeow) [inline code/game/objects/items/devices/flashlight.dm:130]: No `</span>`

File: code/game/objects/items/devices/flashlight.dm [hunk 9]
```diff
@@ -17,18 +25,84 @@
...
+	if(uses_battery)
+		if(istype(int_battery))
+			. += "The flashlight is at [round(int_battery.percent() )]%."
+			. += "<span class='notice'>Alt-click to remove cell.</span>"
+		else
+			. += "There doesn't seem to be a cell inserted."
 
+/obj/item/flashlight/Destroy()
+	if(on)
+		STOP_PROCESSING(SSobj, src)
```
Comment (itsmeow) [inline code/game/objects/items/devices/flashlight.dm:49]: QDEL_NULL the battery please

**Comment by CydiaLamiales** [comment]
Will you also change seclites, penlights and desk lights to require a battery? Also how does this affect the traitor "emp flashlight" item?

File: code/game/objects/items/devices/flashlight.dm [hunk 10]
```diff
@@ -1,3 +1,11 @@
+#define LIGHT_POWER_HIGH 3
+
+#define LIGHT_POWER_LOW 2
+#define LIGHT_CHARGE_LOW 400
+
+#define LIGHT_POWER_CRIT 1
```
Comment (itsmeow) [inline code/game/objects/items/devices/flashlight.dm:6]: `LIGHT_BRIGHTNESS_X` is probably less confusing

**Review by itsmeow — CHANGES_REQUESTED** [review]
Any subtypes of flashlight will also have these changes.

    Flares
    Penlights
    Lamps
    Seclites
    Spotlights
    Glowsticks
    Eyelights

This is bad - you need use_battery to FALSE on them.

Also seclite is a big balance change for security

File: code/game/objects/items/devices/flashlight.dm (hunk 5 above)
Comment (PowerfulBacon) [reply code/game/objects/items/devices/flashlight.dm:42]: Examine doesn't need span

File: code/game/objects/items/devices/flashlight.dm (hunk 8 above)
Comment (EvilDragonfiend) [reply code/game/objects/items/devices/flashlight.dm:130]: Interpretation: no `</span>` means you forgot to put.

File: code/game/objects/items/devices/flashlight.dm [hunk 11]
```diff
@@ -17,18 +25,84 @@
...
+				if(LIGHT_POWER_CRIT)
+					light_range = initial(light_range) / 4
+				else
+					to_chat(user, "\The [src] flickers weakly, but doesn't have enough power to turn on.")
+					on = FALSE
+					return .()
+
+			icon_state = "[initial(icon_state)]-on"
+			playsound(src, 'sound/items/flashlight_on.ogg', 25, 1)
+			START_PROCESSING(SSobj, src)
```
Comment (PowerfulBacon) [inline code/game/objects/items/devices/flashlight.dm:105]: You start processing in here, but update_brightness is called from process()

File: code/game/objects/items/devices/flashlight.dm [hunk 12]
```diff
@@ -17,18 +25,84 @@
...
 	update_brightness()
 
+/obj/item/flashlight/process(delta_time)
+	if(!uses_battery)
+		STOP_PROCESSING(SSobj, src)
+		return
+	if(int_battery)
+		int_battery.use(battery_use_rate * delta_time)
+	update_power()
+	update_brightness()
```
Comment (PowerfulBacon) [inline code/game/objects/items/devices/flashlight.dm:70]: Updates should only happen when needed

File: code/game/objects/items/devices/flashlight.dm [hunk 13]
```diff
@@ -45,6 +119,23 @@
 		A.UpdateButtonIcon()
 	return 1
 
+/obj/item/flashlight/attackby(obj/item/I, mob/user)
```
Comment (PowerfulBacon) [inline code/game/objects/items/devices/flashlight.dm:122]: This needs to return ..() if its not a cell, or it probably makes them impossible to attack

File: code/game/objects/items/devices/flashlight.dm [hunk 14]
```diff
@@ -17,18 +25,84 @@
...
-		icon_state = "[initial(icon_state)]-on"
-		playsound(src, 'sound/items/flashlight_on.ogg', 25, 1)
+		if(int_battery)
+			switch(battery_power_amt)
+				if(LIGHT_POWER_HIGH)
+					light_range = initial(light_range)
+				if(LIGHT_POWER_LOW)
+					light_range = initial(light_range) / 2
+				if(LIGHT_POWER_CRIT)
+					light_range = initial(light_range) / 4
```
Comment (PowerfulBacon) [inline code/game/objects/items/devices/flashlight.dm:97]: Does this work for overlay lights?

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]

**Pull request closed without merging**